ipld/get_shares.go

```diff
@@ -0,0 +1,101 @@
+package ipld
+
+import (
+	"context"
+	"sync"
+
+	"github.com/gammazero/workerpool"
+	"github.com/ipfs/go-cid"
+	format "github.com/ipfs/go-ipld-format"
+)
+
+// NumWorkersLimit sets global limit for workers spawned by GetShares.
+// GetShares could be called MaxSquareSize(128) times per data square each spawning up to 128/2 goroutines and
+// altogether this is 8192. Considering there can be N blocks fetched at the same time, e.g. during catching up data
+// from the past, we multiply this number by the amount of allowed concurrent data square fetches(NumConcurrentSquares).
+// NOTE: This value only limits amount of simultaneously running workers that are spawned as the load increases and
+// are killed, once the load declines. Thanks to simple yet powerful workerpool library by @gammazero.
+// TODO(@Wondertan): This assumes we have parallelized DASer implemented. Sync the values once it is shipped.
+// TODO(@Wondertan): Allow configuration of values.
+var NumWorkersLimit = MaxSquareSize * MaxSquareSize / 2 * NumConcurrentSquares
+
+// NumConcurrentSquares limits the amount of squares that are fetched concurrently/simultaneously.
+var NumConcurrentSquares = 8
+
+// Global worker pool that globally controls and limits goroutines spawned by GetShares.
+var pool = workerpool.New(NumWorkersLimit)
+
+// GetShares gets shares from either local storage, or, if not found, requests them from immediate/connected peers.
+// It puts them into the given 'put' func, does not return any error, and returns/unblocks only on success
+// (got all shares) or on context cancellation.
+//
+// It works concurrently by spawning workers in the pool which do one basic thing - block until data is fetched, s. t.
+// share processing is never sequential, and thus we request *all* the shares available without waiting for others to
+// finish. It is the required property to maximize data availability. As a side effect, we get concurrent tree traversal
+// reducing time to data time.
+//
+// GetShares heavily abuses the fact that the underlying data structure is a binary tree (bin-tree-feat),
```

Comment:
bin-tree-feat?

Reply:
Look at the implementation comments. I've pointed out parts of the code that relies on binary tree properties with this tag

Reply:
Ah I see. Then we should change the comment slightly: https://github.com/celestiaorg/celestia-node/pull/682/files#r869729453